api/app/src/main/kotlin/packit/controllers/PacketController.kt

```diff
@@ -77,4 +81,29 @@ class PacketController(
             .headers(response.second)
             .body(response.first)
     }
+
+    @GetMapping("/{name}/{id}/zip")
+    @PreAuthorize("@authz.canReadPacket(#root, #id, #name)")
+    // TODO: Verify authorization for accessing specific files: whether user has access to all files, or artefacts only.
+    fun downloadZip(
+        @PathVariable id: String,
+        @PathVariable name: String,
+        @RequestParam(required = true) hashes: List<String>,
+        @RequestParam(required = true) filenames: List<String>,
+    ): ResponseEntity<StreamingResponseBody>
+    {
+        val filesWithNames = hashes.mapIndexed { index, hash ->
```

Comment:
I would use a [`zip`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/zip.html) (unfortunate naming collision) here rather than indexing, allowing you to iterate over two lists at the same time.

Something like:
```kotlin
val filesWithNames = hashes.zip(filenames) { hash, filename ->
    val fileResource = packetService.getFileByHash(hash, false, hash).first
    fileResource to filename
}
```

Alternatively make this a POST and put a JSON "filename -> hash" map in the request body.